extends Node

@export var snake_scene : PackedScene

var score : int
var game_started : bool = false

var cells : int = 20
var cells_size : int = 50

var food_pos: Vector2
var regen_food : bool = true

var old_data : Array
var snake_data : Array
var snake : Array

var start_pos = Vector2(9, 9)
var up = Vector2(0, -1)
var down = Vector2(0, 1)
var left = Vector2(-1, 0)
var right = Vector2(1, 0)
var move_direction : Vector2
var can_move: bool



func _ready() -> void:
	new_game()

func new_game():
	score = 0
	$Hud.get_node("ScoreLabel").text = "SCORE: " + str(score)
	move_direction = up
	can_move = true
	generate_snake()
	move_food()

func generate_snake():
	old_data.clear()
	snake_data.clear()
	snake.clear()
	for i in range(3):
		add_segment(start_pos + Vector2(0, i))

func add_segment(pos):
	snake_data.append(pos)
	var SnakeSegment = snake_scene.instantiate()
	SnakeSegment.position = (pos * cells_size) + Vector2(0, cells_size)
	add_child(SnakeSegment)
	snake.append(SnakeSegment)

func _process(delta):
	move_snake()

func move_snake():
	if can_move:
		if Input.is_action_just_pressed("move_down") and move_direction != up:
			move_direction = down
			can_move = false
			if not game_started:
				game_start()
		if Input.is_action_just_pressed("move_up") and move_direction != down:
			move_direction = up
			can_move = false
			if not game_started:
				game_start()
		if Input.is_action_just_pressed("move_right") and move_direction != left:
			move_direction = right
			can_move = false
			if not game_started:
				game_start()
		if Input.is_action_just_pressed("move_left") and move_direction != right:
			move_direction = left
			can_move = false
			if not game_started:
				game_start()

func game_start():
	game_started = true
	$MoveTimer.start()


func _on_move_timer_timeout():
	can_move = true
	old_data = [] + snake_data
	snake_data[0] += move_direction
	for i in range(len(snake_data)):
		if (i > 0):
			snake_data[i] = old_data[i - 1]
		snake[i].position = (snake_data[i] * cells_size) + Vector2(0, cells_size)
	check_out_of_bound()
	check_self_eaten()
	check_food_eaten()

func check_out_of_bound():
	if snake_data[0].x < 0 or snake_data[0].x > cells - 1 or snake_data[0].y < 0 or snake_data[0].y > cells - 1:
		end_game()

func check_self_eaten():
	for i in range (1, len(snake_data)):
		if snake_data[0] == snake_data[i]:
			end_game()

func end_game():
	pass

func check_food_eaten():
	if snake_data[0] == food_pos:
		score += 1
		$Hud.get_node("ScoreLabel").text = "Score: " + str(score)
		add_segment(old_data[-1])
		move_food()

func move_food():
	while regen_food:
		regen_food = false
		food_pos = Vector2(randi_range(0, cells - 1), randi_range(0, cells - 1))
		for i in snake_data:
			if food_pos == i:
				regen_food = true
	$Food.position = (food_pos * cells_size) + Vector2(0, cells_size)
	regen_food = true
